File: rust/crates/core/tests/project_table_store_manifest.rs
use std::time::{SystemTime, UNIX_EPOCH};

use streamfind_rust_core::{Project, ProjectOptions, ProjectTableStore};

fn project() -> (Project, std::path::PathBuf) {
    let stamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    let path = streamfind_rust_test_support::tmp_projects_dir()
        .join(format!("manifest-proof-{stamp}.duckdb"));
    let project = Project::create(ProjectOptions {
        database_path: path.clone(),
        domain: "mass_spec".into(),
        create_if_missing: false,
        read_only: false,
    })
    .unwrap();
    (project, path)
}

fn sql_type(kind: &str) -> &'static str {
    match kind.rsplit('#').next().unwrap_or(kind) {
        "boolean" => "BOOLEAN",
        "integer" => "INTEGER",
        "real" => "DOUBLE",
        "timestamp" => "TIMESTAMP",
        _ => "VARCHAR",
    }
}

fn install(project: &mut Project, omit_table: bool, omit_column: bool, wrong_type: bool) {
    let manifest = streamfind_rust_core::catalogue::table_manifest("mass_spec").unwrap();
    let first_table = manifest
        .iter()
        .find(|table| table.0.starts_with("MASS_SPEC_"))
        .map(|table| table.0.clone());
    let first_column = manifest
        .iter()
        .find(|table| table.0.starts_with("MASS_SPEC_"))
        .and_then(|table| table.1.first().map(|column| column.0.clone()));
    for (table, columns) in manifest {
        if omit_table && Some(table.clone()) == first_table {
            continue;
        }
        let definitions = columns
            .into_iter()
            .filter(|(name, _)| !(omit_column && Some(name.clone()) == first_column))
            .map(|(name, kind)| {
                let ty = if wrong_type && Some(name.clone()) == first_column {
                    "INTEGER"
                } else {
                    sql_type(&kind)
                };
                format!("\"{name}\" {ty}")
            })
            .collect::<Vec<_>>();
        project
            .execute_sql(&format!(
                "CREATE TABLE IF NOT EXISTS \"{table}\" ({})",
                definitions.join(", ")
            ))
            .unwrap();
    }
}

#[test]
fn valid_generated_manifest_is_accepted() {
    let (mut project, path) = project();
    install(&mut project, false, false, false);
    ProjectTableStore::new(&project)
        .require_manifest("mass_spec")
        .unwrap();
    drop(project);
    std::fs::remove_file(path).unwrap();
}

#[test]
fn missing_table_is_rejected() {
    let (mut project, path) = project();
    install(&mut project, true, false, false);
    assert!(ProjectTableStore::new(&project)
        .require_manifest("mass_spec")
        .is_err());
    drop(project);
    std::fs::remove_file(path).unwrap();
}

#[test]
fn missing_column_is_rejected() {
    let (mut project, path) = project();
    install(&mut project, false, true, false);
    assert!(ProjectTableStore::new(&project)
        .require_manifest("mass_spec")
        .is_err());
    drop(project);
    std::fs::remove_file(path).unwrap();
}

#[test]
fn incompatible_column_type_is_rejected() {
    let (mut project, path) = project();
    install(&mut project, false, false, true);
    assert!(ProjectTableStore::new(&project)
        .require_manifest("mass_spec")
        .is_err());
    drop(project);
    std::fs::remove_file(path).unwrap();
}

File: cpp/tests/unit/project_table_store_manifest.cpp
#include <cassert>
#include <array>
#include <filesystem>
#include <string>
#include <tuple>
#include <iostream>

#include "streamfind/catalogue.hpp"
#include "streamfind/project.hpp"
#include "streamfind/project_table_store.hpp"
#include "../tmp_projects.hpp"

namespace streamfind::test::manifest_proof {
std::string sql_type(const std::string &kind) {
    const auto hash = kind.rfind('#');
    const auto name = kind.substr(hash == std::string::npos ? 0 : hash + 1);
    if (name == "boolean") return "BOOLEAN";
    if (name == "integer") return "INTEGER";
    if (name == "real") return "DOUBLE";
    if (name == "timestamp") return "TIMESTAMP";
    return "VARCHAR";
}

void install(streamfind::Project &project, bool omit_table, bool omit_column, bool wrong_type) {
    const auto manifest = streamfind::catalogue::table_manifest_json("mass_spec", "mass_spec.base");
    assert(manifest);
    std::string first_table;
    std::string first_column;
    for (const auto &table : *manifest) {
        if (table.at("table_name").get<std::string>().rfind("MASS_SPEC_", 0) == 0) {
            first_table = table.at("table_name");
            if (!table.at("columns").empty()) first_column = table.at("columns").front().at("name");
            break;
        }
    }
    for (const auto &table : *manifest) {
        const auto name = table.at("table_name").get<std::string>();
        if (omit_table && name == first_table) continue;
        std::string ddl = "CREATE TABLE IF NOT EXISTS \"" + name + "\" (";
        bool first = true;
        for (const auto &column : table.at("columns")) {
            const auto column_name = column.at("name").get<std::string>();
            if (omit_column && column_name == first_column && name == first_table) continue;
            if (!first) ddl += ", ";
            first = false;
            const auto type = wrong_type && column_name == first_column && name == first_table
                ? "INTEGER" : sql_type(column.at("type").get<std::string>());
            ddl += "\"" + column_name + "\" " + type;
        }
        ddl += ")";
        project.execute_sql(ddl);
    }
}

int run() {
    const auto base = streamfind::test::tmp_projects_dir();
    const std::array cases{
        std::tuple{"valid", false, false, false},
        std::tuple{"missing-table", true, false, false},
        std::tuple{"missing-column", false, true, false},
        std::tuple{"wrong-type", false, false, true},
    };
    for (const auto &[suffix, omit_table, omit_column, wrong_type] : cases) {
        const auto path = base / (std::string("manifest-proof-") + suffix + ".duckdb");
        std::error_code error;
        std::filesystem::remove(path, error);
        auto project = streamfind::Project::create({path, "mass_spec", {}});
        install(project, omit_table, omit_column, wrong_type);
        bool passed = true;
        try {
            streamfind::ProjectTableStore(project).require_manifest("mass_spec", "mass_spec.base");
        } catch (const streamfind::Error &exception) {
            std::cerr << suffix << ": " << exception.what() << '\n';
            passed = false;
        }
        if (suffix == std::string("valid") ? !passed : passed) return 1;
        project.close();
        std::filesystem::remove(path, error);
    }
    return 0;
}
} // namespace streamfind::test::manifest_proof

int main() { return streamfind::test::manifest_proof::run(); }
